### PR TITLE
Don't start config thread until _after_ components are set up

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim please review
@hakonhall FYI

Avoids race where new config received after initial fetch but before node is fully set up gets lost.

### CODE REVIEW

**Review by baldersheim — APPROVED** [review]
I also seem remember an if where handleLiveC.. was called from that could either be an assert or be removed.

**Comment by vekterli** [comment]
That would be the check for whether the `configure()` call is the first (synchronous) subscription callback, or a subsequent asynchronous reconfig. This should still work, but now without the potential of happening before the components are set up.

**Comment by vekterli** [comment]
storageserver tests are complaining, will check it out

**Comment by vekterli** [comment]
@baldersheim please take another look.

Semantics of config subscriptions are different than I originally thought, i.e. no sync callback upon `subscribe()`. Callbacks are not invoked until `ConfigFetcher::start()` is called, after which point async callbacks may happen at any time. Redesigned solution in the pragmatic way most problems are handled in multiprogramming: add another mutex! 🎉

File: storage/src/vespa/storage/storageserver/storagenode.cpp [hunk 1]
```diff
@@ -247,11 +252,10 @@ StorageNode::updateUpgradeFlag(const vespa::config::content::UpgradingConfig& co
 void
 StorageNode::handleLiveConfigUpdate()
 {
```
Comment (baldersheim) [inline storage/src/vespa/storage/storageserver/storagenode.cpp]: I think it wise signal the lock must be held in the signature.

**Comment by baldersheim** [comment]
@vekterli I added a pull request with the guard in the signature. I caught usage from ServiceLayerNode that I think needed the same guard.

**Comment by baldersheim** [comment]
from my point of view this is now ready to go.

File: storage/src/vespa/storage/storageserver/storagenode.cpp (hunk 1 above)
Comment (vekterli) [reply storage/src/vespa/storage/storageserver/storagenode.cpp]: Agreed. The less we allow gun vs. foot interaction, the better.

**Pull request merged**